packages/@eventual/aws-runtime/src/stores/bucket-store.ts

```diff
@@ -0,0 +1,217 @@
+} from "@eventual/core-runtime";
+import { assertNever } from "@eventual/core/internal";
+import { Readable } from "stream";
+import { bucketServiceBucketName } from "../utils.js";
+
+export interface BucketRuntimeOverrides {
+  /**
+   * Override the s3 bucket name of the bucket.
+   */
+  bucketName?: string;
+}
+
+export interface AWSBucketStoreProps {
+  s3: S3Client;
+  serviceName: LazyValue<string>;
+  bucketOverrides: LazyValue<Record<string, BucketRuntimeOverrides>>;
+}
+
+export class AWSBucketStore implements BucketStore {
+  constructor(private props: AWSBucketStoreProps) {}
+  public async get(
+    bucketName: string,
+    key: string,
+    options?: GetBucketObjectOptions
+  ): Promise<GetBucketObjectResponse | undefined> {
+    const result = await this.props.s3.send(
+      new GetObjectCommand({
+        Bucket: this.physicalName(bucketName),
+        Key: key,
+        IfMatch: options?.etag,
+      })
+    );
+
+    return result.Body
+      ? {
+          body: result.Body as Readable,
+          contentLength: result.ContentLength!,
+          etag: result.ETag,
+        }
+      : undefined;
+  }
+
+  public async head(
+    bucketName: string,
+    key: string,
+    options?: GetBucketObjectOptions
+  ): Promise<GetBucketMetadataResponse | undefined> {
+    const result = await this.props.s3.send(
+      new HeadObjectCommand({
+        Bucket: this.physicalName(bucketName),
+        Key: key,
+        IfMatch: options?.etag,
+      })
+    );
+
+    return result.ContentLength !== undefined
+      ? {
+          contentLength: result.ContentLength,
+          etag: result.ETag,
+        }
+      : undefined;
+  }
+
+  public async put(
+    bucketName: string,
+    key: string,
+    data: string | Readable | Buffer
+  ): Promise<PutBucketObjectResponse> {
+    const result = await this.props.s3.send(
+      new PutObjectCommand({
+        Bucket: this.physicalName(bucketName),
+        Key: key,
+        Body: data,
+      })
+    );
+
+    return {
+      etag: result.ETag,
+    };
+  }
+
+  public async delete(bucketName: string, key: string): Promise<void> {
+    await this.props.s3.send(
+      new DeleteObjectCommand({
+        Bucket: this.physicalName(bucketName),
+        Key: key,
+      })
+    );
+  }
+
+  public async copyTo(
+    bucketName: string,
+    key: string,
+    sourceKey: string,
+    sourceBucket?: Bucket | undefined,
+    options?: CopyBucketObjectOptions | undefined
+  ): Promise<CopyBucketObjectResponse> {
+    const result = await this.props.s3.send(
+      new CopyObjectCommand({
+        Bucket: this.physicalName(bucketName),
+        Key: key,
+        CopySource: `${
+          sourceBucket
+            ? this.physicalName(sourceBucket.name)
+            : this.physicalName(bucketName)
+        }/${sourceKey}`,
+        CopySourceIfMatch: options?.sourceEtag,
+      })
+    );
+
+    return {
+      etag: result.CopyObjectResult?.ETag,
+    };
+  }
+
+  public async presignedUrl(
```

Comment:
This should be named `getPresignedUrl` or `presignUrl`

Reply:
HMMMMMMMMM, I don't like `getPresignedUrl` as the presigned url need to be given an operation where `getPresignedUrl("get")` and `getPresignedUrl("put")` are kind of odd.

AWS uses `S3RequestPresigner.signUrl` an `getSignedUrl` in the SDK.

How about `generatePresignedUrl`?

Reply:
Updated